group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/SubscribedTopicDescriberImpl.java

```diff
@@ -66,22 +64,18 @@ public int numPartitions(Uuid topicId) {
      */
     @Override
     public Set<String> racksForPartition(Uuid topicId, int partition) {
-        TopicImage topic = metadataImage.topics().getTopic(topicId);
-        if (topic != null) {
-            PartitionRegistration partitionRegistration = topic.partitions().get(partition);
-            if (partitionRegistration != null) {
-                Set<String> racks = new HashSet<>();
-                for (int replica : partitionRegistration.replicas) {
-                    // Only add the rack if it is available for the broker/replica.
-                    BrokerRegistration brokerRegistration = metadataImage.cluster().broker(replica);
-                    if (brokerRegistration != null) {
-                        brokerRegistration.rack().ifPresent(racks::add);
-                    }
-                }
-                return Collections.unmodifiableSet(racks);
-            }
+        Optional<CoordinatorMetadataImage.TopicMetadata> topicMetadataOp = metadataImage.topicMetadata(topicId);
+        if (topicMetadataOp.isEmpty()) {
+            return Collections.emptySet();
+        }
+
+        CoordinatorMetadataImage.TopicMetadata topicMetadata = topicMetadataOp.get();
+        List<String> racks = topicMetadata.partitionRacks().get(partition);
```

Comment:
This call is pretty nasty because it generate the entire map just to return the racks of a single partition and it does so every time racksForPartition is called. We should definitely consider my suggestion for the interface.